editor/src/clj/internal/graph.clj

```diff
@@ -813,50 +813,40 @@
-           todo (seq outputs-by-node-ids)
-           ;; collect next batch of entries in a map, to coalesce common node ids
-           next-todo {}
-           ;; final transitive closure of entries found, as a map
-           result {}]
-      (if-let [[node-id outputs] (first todo)]
-        (let [seen? (get result node-id)]
-          ;; termination condition is when we have seen *every* output already
-          (if (and seen? (every? seen? outputs))
-            ;; completely remove the node-id from todo as we have seen *every* output
-            (recur (next todo) next-todo result)
-            ;; does the node-id have any successors at all?
-            (if-let [label->succ (node-id->successor-map node-id)]
-              ;; ignore the outputs we have already seen
-              (let [outputs (if seen? (set/difference outputs seen?) outputs)
-                    ;; Add every successor to the stack for later processing
-                    next-todo (transduce (map label->succ) merge-with-union next-todo outputs)
-                    ;; And include the unseen output labels to the result
-                    result (update result node-id set-or-union outputs)]
-                (recur (next todo) next-todo result))
-              ;; There were no successors, recur without that node-id
-              (recur (next todo) next-todo result))))
-        ;; check if there is a next batch of entries to process
-        (if-let [todo (seq next-todo)]
-          (recur todo {} result)
-          result)))))
-
+(defn- basis-dependencies [basis node-id+outputs]
+  (let [graph-id->node-successor-map
+        (persistent!
+          (reduce-kv
+            (fn [acc graph-id graph]
+              (assoc! acc graph-id (:successors graph)))
+            (transient {})
+            (:graphs basis)))
+        node-id+output->successors (fn [node-id+output]
+                                     (let [node-id (node-id+output 0)
+                                           output (node-id+output 1)]
+                                       (-> node-id
+                                           gt/node-id->graph-id
+                                           graph-id->node-successor-map
+                                           (get node-id)
+                                           (get output))))
+        deque (ArrayDeque.)
+        result (HashSet.)]
+    (reduce (fn [_ node-id+output]
+              (.push deque node-id+output))
+            nil
+            node-id+outputs)
+    (loop []
+      (when-some [node-id+output (.pollLast deque)]
+        (when-not (.contains result node-id+output)
+          (when-some [successors (node-id+output->successors node-id+output)]
+            (reduce
+              (fn [_ successor-node-id+output]
+                (.push deque successor-node-id+output))
+              nil
+              successors)
+            (.add result node-id+output)))
+        (recur)))
+    result))
```

Comment:
I'm not sure I can tell if the new implementation achieves the same exact result as the earlier implementation. 😬 Do we have good test coverage of this?